File: help_functions/chatgpt.lua
local function gather_response(prompt, api_key)
    local endpoint = "https://api.openai.com/v1/chat/completions"

    local headers = {
        ["Authorization"] = "Bearer " .. api_key,
        ["Content-Type"] = "application/json"
    }

    local payloadJson = textutils.serializeJSON({
        ["model"] = "gpt-3.5-turbo",
        ["max_tokens"] = 100,
        ["messages"] = {{
                role = "user",
                content = prompt
            }},
    })

    -- Send the HTTP request to the OpenAI API
    local requesting = true
    http.request(endpoint, payloadJson, headers)

    -- Wait for HTTP response 
    while requesting do
        local event, url, sourceText = os.pullEvent()

        -- Handle successful HTTP response
        if event == "http_success" then
            local responseBody = sourceText.readAll()
            sourceText.close()
            local responseJson = textutils.unserializeJSON(responseBody)

            -- Check if a valid response was received
            if responseJson and responseJson.choices and responseJson.choices[1] then
                local reply = responseJson.choices[1].message.content

                -- Print the generated response from the model
                print("Response: " .. reply)
                return reply
            else
                print("Error: Failed to get a valid response.")
            end

            requesting = false

            -- Handle failed HTTP response
        elseif event == "http_failure" then
            print("Server didn't respond.")
            requesting = false
        end
    end
end

return { gather_response = gather_response }

File: help_functions/chatgpt.py
import openai
import sys
import os
import time

def check_file():
    try:
        f = open("prompt.txt", "r")
        prompt = f.read()
        print(f'removing prompt.txt')
        os.remove("prompt.txt")

        return prompt
    except:
        return ""

def chatgpt_response(prompt, api_key):
    openai.api_key = api_key
    messages = [ {"role": "system", "content": prompt} ]

    chat = openai.ChatCompletion.create( 
        model = "gpt-3.5-turbo", messages = messages
    )

    reply = chat.choices[0].message.content

    return reply

def save_to_file(reply):
    f = open("reply.txt", "w")
    print(f'writing reply.txt')
    f.write(reply)
    f.close()


if __name__ == "__main__":
    prompt = ""
    api_key = open("openai.key", "r").read()

    while True:
        while prompt == "":
            prompt = check_file()
            time.sleep(1)

        api_key = sys.argv[1]
        reply = chatgpt_response(prompt, api_key)

        save_to_file(reply)
